Don't rely on hypothesis internals in test suite

<!-- Thanks for reporting a bug for this project. READ THIS FIRST:

- Provide as many details as possible. Simply saying "X gives bug" or "X gives error" is not enough!
- Paste logs, configuration sample and code into the backticks (```).
- Read through the template carefully and fill out all missing details.
- Please also search for similar issues in this issue tracker first and read through the ESPHome FAQ.

DO NOT DELETE ANY TEXT from this template! Otherwise the issue may be closed without a comment.
-->

**Operating environment/Installation (Hass.io/Docker/pip/etc.):** NixOS
<!--
Please provide details about your environment below this line. --> 

We are a linux distribution packaging esphome. This means we have to relax some of esphome's version pins. Our pytest session looks like this:

```
platform linux -- Python 3.8.9, pytest-6.2.3, py-1.10.0, pluggy-0.13.1 -- /nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/bin/python3.8
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/build/source/.hypothesis/examples')
rootdir: /build/source, configfile: pytest.ini
plugins: cov-2.11.1, mock-3.5.1, hypothesis-5.49.0
collected 299 items / 1 error / 298 selected
```

**ESP (ESP32/ESP8266, Board/Sonoff):**
<!--
Please provide details about which ESP you're using below.
-->
n/a

**ESPHome version (latest production, beta, dev branch)**
<!--
ESPHome version like v1.14 or 1.15-dev
-->
1.17.0

**Affected component:**
<!--
Please add the link to the documentation at https://esphome.io/index.html of the component in question.
-->
Test suite.

**Description of problem:**

The test suite relies on internals of hypothesis, that have no stability guarantees and are clearly marked internal:

```

________________ ERROR collecting tests/unit_tests/test_core.py ________________
tests/unit_tests/test_core.py:5: in <module>
    from strategies import mac_addr_strings
tests/unit_tests/strategies.py:7: in <module>
    @st.defines_strategy_with_reusable_values
E   AttributeError: module 'hypothesis.strategies._internal.core' has no attribute 'defines_strategy_with_reusable_values'
```

**Problem-relevant YAML-configuration entries:**
n/a

**Logs (if applicable):**
<!--
Please copy the debug log here. If possible, also connect to the ESP over USB and copy those logs into the backticks.
-->
n/a

**Additional information and things you've tried:**

```diff
-@st.defines_strategy_with_reusable_values
+@st.defines_strategy(force_reusable_values=True)
```

Making this substitution fixes the test suite with hypothesis>=5.32.1.

https://github.com/HypothesisWorks/hypothesis/commit/a65aff6e03398e37a96b77d8a89a4f5591f5cf8b

<!-- LEAVE THIS LINE AS-IS AND DON'T DELETE IT, OTHERWISE THE ISSUE WILL BE CLOSED AUTOMATICALLY. -->

--- a/requirements_test.txt
+++ b/requirements_test.txt
@@ -11,4 +11,4 @@ pytest==6.2.3
 pytest-cov==2.11.1
 pytest-mock==3.5.1
 asyncmock==0.4.2
-hypothesis==5.21.0
+hypothesis==5.49.0

--- a/tests/unit_tests/strategies.py
+++ b/tests/unit_tests/strategies.py
@@ -4,7 +4,7 @@ import hypothesis.strategies._internal.core as st
 from hypothesis.strategies._internal.strategies import SearchStrategy
 
 
-@st.defines_strategy_with_reusable_values
+@st.defines_strategy(force_reusable_values=True)
 def mac_addr_strings():
     # type: () -> SearchStrategy[Text]
     """A strategy for MAC address strings.

